File: src/main/java/frc/robot/Robot.java
package frc.robot;

import java.util.ArrayList;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.Set;

import org.team997coders.spartanlib.controllers.SpartanPID;
import org.team997coders.spartanlib.helpers.PIDConstants;
import org.team997coders.spartanlib.helpers.SwerveMixerData;
import org.team997coders.spartanlib.helpers.threading.SpartanRunner;
import org.team997coders.spartanlib.limelight.LimeLight;
import org.team997coders.spartanlib.motion.pathfollower.PathManager;
import org.team997coders.spartanlib.motion.pathfollower.data.PathData;

import edu.wpi.first.networktables.NetworkTable;
import edu.wpi.first.networktables.NetworkTableEntry;
import edu.wpi.first.networktables.NetworkTableInstance;
import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj2.command.*;
import edu.wpi.first.wpilibj.smartdashboard.*;
import frc.robot.commands.auto.*;
import frc.robot.commands.drivetrain.*;
import frc.robot.commands.hopper.*;

import frc.robot.commands.shooter.ShootBadly;
import frc.robot.subsystems.*;

public class Robot extends TimedRobot {

  public static SpartanRunner mRunner = new SpartanRunner(20);

  private static double lastUpdate = 0.0;
  public static double initAngle = 0.0;
  private ArrayList<String> commandList;

  public static long cycles = 0;
  public static final boolean verbose = false; //debug variable, set to true for command logging in the console and non-essential smartdashboard bits.

  private Command m_autonomousCommand;
  private Command mHopperCommand;
  public static boolean autoLoadHopper = false;

  Command autonomousCommand;
  SendableChooser<Command> m_chooser = new SendableChooser<>();
  
  @Override
  public void robotInit() {

    // CameraServer.getInstance().startAutomaticCapture(0);
    requestPaths();

    LimeLight.getInstance().setDouble(LimeLight.LED_MODE, LimeLight.LEDState.ForceOff);

    CommandScheduler.getInstance().cancelAll();

    m_chooser.setDefaultOption("Do Nothing", new AutoDoNothing());
    m_chooser.addOption("Auto One", new FollowPath("TrenchPivot", false));
    m_chooser.addOption("s1ck0 m0d3", new AutoSickoMode());
    m_chooser.addOption("Shoot Balls", new AutoStreamUntilEmpty(Constants.Values.SHOOTER_RPM, true));
    m_chooser.addOption("Drive Forward", new FollowPath("DriveForward", false));
    m_chooser.addOption("AutoNav One", new FollowPath("BarrelRacing", false));
    m_chooser.addOption("AutoNav Two", new FollowPath("Slalom", false));
    m_chooser.addOption("AutoNav Three", new FollowPath("Bounce", false));

    LimeLight.getInstance().mController = new SpartanPID(new PIDConstants(
      Constants.Values.VISION_TURNING_P,
      Constants.Values.VISION_TURNING_I,
      Constants.Values.VISION_TURNING_D
    ));

    Shooter.getInstance();
    Hopper.getInstance();
    DriveTrain.getInstance();
    DriveTrain.getInstance().setDefaultCommand(new SwerveMixer());
    Climber.getInstance();

    LEDManager.getInstance();

    OI.getInstance();

    commandList = new ArrayList<String>();
    if (verbose) {
      CommandScheduler.getInstance().onCommandExecute(command -> {
        commandList.add(command.getName());
      });
    }

    SmartDashboard.putData(m_chooser);
    SmartDashboard.putNumber("Driver/Set Initial Angle", 0.0); // set the init angle of the robot in disabled with this. 0 is straight forwards.

    mHopperCommand = new HopperAutoIndex();
  }

  @Override
  public void robotPeriodic() {
    
    if (verbose) {
      if (commandList.size() > 0) {
        for (String cmdName : commandList) {
          System.out.println("Ran " + cmdName + " on cycle " + cycles);
        }
      }
    }
    commandList.clear();
    cycles++;

    lastUpdate = getCurrentSeconds();

    OI.getInstance().update();
  }

  @Override
  public void disabledInit() {

    DriveTrain.getInstance().setCoast();

    LimeLight.getInstance().setDouble(LimeLight.LED_MODE, LimeLight.LEDState.ForceOff);
    Intake.getInstance().setPiston(false);

    if (mHopperCommand != null) mHopperCommand.cancel();
    mHopperCommand = new HopperAutoIndex();

    LEDManager.getInstance().setColor(1, 0, 100);

  }

  @Override
  public void disabledPeriodic() {
    CommandScheduler.getInstance().run();

    Hopper.getInstance().updateBallCount();
    initAngle = NetworkTableInstance.getDefault().getTable("SmartDashboard").getEntry("Driver/Set Init Angle").getDouble(0.0);
  }

  @Override
  public void autonomousInit() {
    m_autonomousCommand = m_chooser.getSelected();

    if (m_autonomousCommand != null) {
      m_autonomousCommand.schedule();
    }
  }

  @Override
  public void autonomousPeriodic() {
    CommandScheduler.getInstance().run();
  }

  @Override
  public void teleopInit() {

    DriveTrain.getInstance().setBrake();

    if (m_autonomousCommand != null) {
      m_autonomousCommand.cancel();
    }

    LEDManager.getInstance().setColorToAlliance();

    mHopperCommand.schedule();
  }

  @Override
  public void teleopPeriodic() {
    CommandScheduler.getInstance().run();
  }

  @Override
  public void testInit() {
    CommandScheduler.getInstance().cancelAll();
  }

  public void requestPaths() {
    for (Map.Entry<String, PathData> entry : Constants.PathFollower.PATH_DATA.entrySet()) {
      PathManager.getInstance().queueData(entry.getValue());
    }
  }

  @Override
  public void testPeriodic() { }

  public static double getDeltaT() { return getCurrentSeconds() - lastUpdate; }

  public static double getCurrentSeconds() { return System.currentTimeMillis() / 1000.0; }

}

File: src/main/kotlin/frc/robot/Robot.kt
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2018 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package frc.robot

import com.ctre.phoenix.motorcontrol.NeutralMode
import edu.wpi.first.wpilibj.TimedRobot
import edu.wpi.first.wpilibj2.command.Command
import edu.wpi.first.wpilibj2.command.CommandScheduler
import frc.robot.commands.drivetrain.ArcadeDrive
import frc.robot.commands.hopper.HopperAutoIndex
import frc.robot.subsystems.Drivetrain
import frc.robot.subsystems.Hopper

/**
 * The VM is configured to automatically run this class, and to call the
 * functions corresponding to each mode, as described in the TimedRobot
 * documentation. If you change the name of this class or the package after
 * creating this project, you must also update the build.gradle file in the
 * project.
 */
class Robot : TimedRobot() {
  /**
   * This function is run when the robot is first started up and should be used
   * for any initialization code.
   */

  companion object {
    var EnableAutoLoader: Boolean = false

    var LastUpdate: Double = 0.0

    var HopperCommand: Command? = null

    fun getCurrentTime(): Double {
      return System.currentTimeMillis() / 1000.0
    }
    fun getDeltaTime(): Double {
      return getCurrentTime() - LastUpdate
    }
  }

  override fun robotInit() {
    Drivetrain.setDefaultCommand(ArcadeDrive())
  }

  override fun robotPeriodic() {
    LastUpdate = getCurrentTime()
  }

  override fun disabledInit() {
    Drivetrain.setNeutralMode(NeutralMode.Coast)

    HopperCommand?.cancel()
    HopperCommand = HopperAutoIndex()
  }

  override fun disabledPeriodic() {
    CommandScheduler.getInstance().run()
  }

  override fun autonomousInit() {
    // TODO: Add auto command
    HopperCommand?.schedule()
  }

  override fun autonomousPeriodic() {
    CommandScheduler.getInstance().run()
  }

  override fun teleopInit() {
    Drivetrain.setNeutralMode(NeutralMode.Brake)
    // TODO: Cancel auto command
    HopperCommand?.schedule()
  }

  override fun teleopPeriodic() {
    CommandScheduler.getInstance().run()
  }

  override fun testInit() {
    CommandScheduler.getInstance().cancelAll()
  }

  override fun testPeriodic() {}
}
